### DISCUSSION TITLE
RTIC app crate that avoids `lto = "thin"`?

### DISCUSSION THREAD

**Discussion by psentee** [mciantyre/teensy4-rs#142] [General]
I'm starting a Teensy4 project, and I want to use RTIC (new shiny 2.0 if it matters), starting with [the `rtic_usb_log` example](https://github.com/mciantyre/teensy4-rs/blob/master/examples/rtic_usb_log.rs), and it works – as long as I set `release` profile to exactly `lto = "thin"` in my `Cargo.toml`. With `lto = false` or `lto = true`/`"fat"` it tails to build with `Linking globals named 'USB_OTG1': symbol multiply defined!`. This happens with both GNU ld and lld.

I found some [cryptic comment](https://github.com/mciantyre/teensy4-rs/blob/master/Cargo.toml#L79-L86) about default features of `teensy4-hal` conflicting with `rtic` and LTO being able to sort it out (in practice, only "thin" LTO). I haven't been able to find any examples of how to set `teensy4-rs` features to work with `rtic`, or any features in its `Cargo.toml` that look like they could do the trick.

I found some historical mentions of patching `cortex-m-rt` in the app to point at `teensy4-rt` crate that doesn't exist nowadays. Also, found some mentions of a replacement/patched `cortex-m-rt` crate that has been removed from this repo in the meantime. I'm confused and have no idea what can I do to make it work other than blindly setting `lto = "thin"` without understanding what I'm doing and giving up on full LTO.

Any hints?

ETA: my code is here: https://github.com/psentee/teensy4-rs-example – it fails to build with `Linking globals named 'USB_OTG1': symbol multiply defined!` and after switching to `lto = "thin"` it builds fine.

**Comment by mciantyre** [comment]
Sorry for the trouble, and thanks for the example code. Could you give https://github.com/psentee/teensy4-rs-example/pull/1 a test?

The BSP's `usb-logging` feature provides a convenience for non-RTIC users. When enabled, the BSP defines `USB_OTG1`. This results in a duplicate definition once an RTIC application also defines `USB_OTG1`. By disabling `usb-logging`, we remove the BSP convenience, allowing users to manage USB logging and `USB_OTG1` however they want.

Let me know if that works. If it does, I'll make sure to note this in the BSP documentation.

**Reply by psentee** [reply]
Besides docs, it might be helpful to add a comment in the `rtic_usb_log` example to explain that this code does not use the `usb-logging` feature – this part was not obvious to me (the code was doing USB logging, so I assumed it was necessary).

**Comment by psentee** [comment]
That fixes the problem, thank you! And I forgot to thank you for all the work on imxrt and teensy4 support! I have hacked together a Rust teensy4 project back in 2019, it was an uphill battle and I can only imagine the amount of work that went into the imxrt-rs/teensy4-rs ecosystem.
